### PR TITLE
Add OpenTelemetry support.

### PR DESCRIPTION
**Pull request by alesj**
Signed-off-by: Ales Justin <ales.justin@gmail.com>

### CODE REVIEW

File: config/application.properties [hunk 1]
```diff
@@ -1,7 +1,8 @@
 #Bridge related settings
 bridge.id=my-bridge
 # uncomment the following line to enable Jaeger tracing, check the documentation how to configure the tracer
-#bridge.tracing=jaeger
+bridge.tracing=jaeger-otel
+bridge.tracing.service-name=strimzi-kafka-bridge
```
Comment (ppatierno) [inline config/application.properties:5]: Shouldn't the above be commented to not have tracing enabled by default? At least it's how works today with the comment above explaining to uncomment to enable tracing.

File: pom.xml [hunk 2]
```diff
@@ -106,6 +106,9 @@
 		<jaeger.version>1.6.0</jaeger.version>
 		<opentracing.version>0.33.0</opentracing.version>
 		<opentracing-kafka-client.version>0.1.15</opentracing-kafka-client.version>
+		<opentelemetry.version>1.7.0-alpha</opentelemetry.version>
+		<opentelemetry-stable.version>1.7.0</opentelemetry-stable.version>
```
Comment (ppatierno) [inline pom.xml:110]: why there are two opentelemetry version properties (one with `-stable` and the other seems to be an alpha) using both in different places to import dependencies?

File: src/main/java/io/strimzi/kafka/bridge/Application.java [hunk 3]
```diff
@@ -62,12 +61,13 @@
     private static final int DEFAULT_EMBEDDED_HTTP_SERVER_PORT = 8080;
 
     @SuppressWarnings({"checkstyle:NPathComplexity"})
+    @SuppressFBWarnings("REC_CATCH_EXCEPTION")
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/Application.java:64]: why this one?

File: src/main/java/io/strimzi/kafka/bridge/MetricsReporter.java [hunk 4]
```diff
@@ -24,6 +25,7 @@
      * @param jmxCollectorRegistry JmxCollectorRegistry instance for scraping metrics from JMX endpoints
      * @param meterRegistry MeterRegistry instance for scraping metrics exposed through Vert.x
      */
+    @SuppressFBWarnings("SIC_INNER_SHOULD_BE_STATIC_ANON")
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/MetricsReporter.java:28]: why this one?

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        return serviceName;
+    }
+
+    @Override
+    public void initialize() {
+        OpenTelemetrySdkAutoConfiguration.initialize();
+    }
+
+    private static Tracer get() {
+        return GlobalOpenTelemetry.getTracer(COMPONENT); // TODO OK?
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:77]: what's this "TODO OK?" ?

File: config/application.properties (hunk 1 above)
Comment (alesj) [reply config/application.properties:5]: Ah, yes, forgot to change it back ... fixed.

File: pom.xml (hunk 2 above)
Comment (alesj) [reply pom.xml:110]: "My" OpenTelemetry artifacts did yet made it to "stable" version, where Jaeger artifcat alfready did.
Hence the 2 diff versions.
But we agreed it's OK to use alpha for those -- as they are the initial ones, and haven't been changed.
I'll check if there is a stable version, but shouldn't make a diff.
(if I change it again now, I need to re-run the whole cycle, for the N-th time ...)

File: src/main/java/io/strimzi/kafka/bridge/Application.java (hunk 3 above)
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java:64]: When I ran `mvn clean install -DskipTests spotbugs:check` it spit out this warning / error ...

File: src/main/java/io/strimzi/kafka/bridge/MetricsReporter.java (hunk 4 above)
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/MetricsReporter.java:28]: Same here, checkstyle / spotbugs complained ...

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java (hunk 5 above)
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:77]: Just an early remark for me, removing it ...

File: src/main/java/io/strimzi/kafka/bridge/Application.java (hunk 3 above)
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/Application.java:64]: this is not the reason, this is the detection of the problem ... I would like to know why there is now a problem that needs this suppress warning that wasn't there before.

File: src/main/java/io/strimzi/kafka/bridge/MetricsReporter.java (hunk 4 above)
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/MetricsReporter.java:28]: ditto as above

File: src/main/java/io/strimzi/kafka/bridge/Application.java (hunk 3 above)
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java:64]: No idea, I was just following spotbugs error report.
Perhaps checkstyle plugin upgrade?

File: src/main/java/io/strimzi/kafka/bridge/MetricsReporter.java (hunk 4 above)
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/MetricsReporter.java:28]: Same ... as you can see, I didn't change anything else in this class.

File: src/main/java/io/strimzi/kafka/bridge/Application.java (hunk 3 above)
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/Application.java:64]: I would try to understand what's the error and then at least opening an issue with the reason to track and doing a change later to avoid doing it in the current PR. If spotbugs is raising this maybe we don't see a potential bug ... it's worth investigating.

Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java:64]: I think it's the issue of "eating" the exception:
* https://github.com/archived-codacy/codacy-findbugs/blob/master/src/main/resources/docs/description/REC_CATCH_EXCEPTION.md

Which in our case that's not the (completely the) case -- we log and exit the app ...

File: pom.xml (hunk 2 above)
Comment (ppatierno) [reply pom.xml:110]: Did you check if there is a stable version? This "alpha" here bothers me a little bit :-)

File: src/main/java/io/strimzi/kafka/bridge/Application.java (hunk 3 above)
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/Application.java:64]: The link you pasted seems to suggest that we should not catch just `Exception` as the main does but `RuntimeException` or being more specific. Would it be the solution instead of suppress the warning?

Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java:64]: Imho, we can leave it as is ... as that's from main() and we log and exit.

File: pom.xml (hunk 2 above)
Comment (alesj) [reply pom.xml:110]: Let me check ...

**Review by ppatierno — APPROVED** [review]

File: pom.xml (hunk 2 above)
Comment (alesj) [reply pom.xml:110]: https://search.maven.org/artifact/io.opentelemetry.instrumentation/opentelemetry-kafka-clients-common

Comment (alesj) [reply pom.xml:110]: All alphas ... but that's OK, since, afaik / imho, the 1st didn't change at all ...

File: src/main/java/io/strimzi/kafka/bridge/tracing/TracingUtil.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    private static TracingHandle tracing = new NoopTracing();
+
+    public static TracingHandle getTracing() {
+        return tracing;
+    }
+
+    public static void initialize(BridgeConfig config) {
+        String tracingConfig = config.getTracing();
+        if (tracingConfig != null && tracingConfig.contains("jaeger")) {
+            boolean isOpenTelemetry = "jaeger-otel".equals(tracingConfig);
```
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/bridge/tracing/TracingUtil.java:31]: Define as `static final String ...` ...

File: src/main/java/io/strimzi/kafka/bridge/tracing/TracingUtil.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    private static final Logger log = LoggerFactory.getLogger(TracingUtil.class);
+    private static TracingHandle tracing = new NoopTracing();
+
+    public static TracingHandle getTracing() {
+        return tracing;
+    }
+
+    public static void initialize(BridgeConfig config) {
+        String tracingConfig = config.getTracing();
+        if (tracingConfig != null && tracingConfig.contains("jaeger")) {
```
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/bridge/tracing/TracingUtil.java:30]: Would it be better to just say `...equals("jaeger") || ...equals("jarger-otel")`?

File: src/main/java/io/strimzi/kafka/bridge/tracing/TracingUtil.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            boolean isOpenTelemetry = "jaeger-otel".equals(tracingConfig);
+            TracingHandle instance = isOpenTelemetry ? new OpenTelemetryHandle() : new OpenTracingHandle();
+
+            String serviceName = instance.serviceName(config);
+            if (serviceName != null) {
+                log.info(
+                    "Initializing Jaeger ({}) tracingConfig with service name {}",
+                    isOpenTelemetry ? "OpenTelemetry" : "OpenTracing",
+                    serviceName
+                );
```
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/bridge/tracing/TracingUtil.java:40]: Does this make sense? Doesn't `serviceName` ecist only for OTel? I.e. would this log line ever print for OpenTelemetry?

File: src/test/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryTest.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+import io.vertx.tracing.opentelemetry.OpenTelemetryOptions;
+
+/**
+ * OpenTelemetry tests
+ */
+public class OpenTelemetryTest extends TracingTestBase {
+    @Override
+    protected TracingOptions tracingOptions() {
+        System.setProperty("otel.traces.exporter", "jaeger");
+        System.setProperty("otel.service.name", "strimzi-kafka-bridge-test");
```
Comment (scholzj) [inline src/test/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryTest.java:18]: This is where you should use the constants defined elsewhere ;-)

**Review by scholzj — COMMENTED** [review]
Did you test the backwards compatibility? I.e. that the old OpenTracing tracing works?

File: config/application.properties [hunk 10]
```diff
@@ -1,7 +1,8 @@
 #Bridge related settings
 bridge.id=my-bridge
 # uncomment the following line to enable Jaeger tracing, check the documentation how to configure the tracer
-#bridge.tracing=jaeger
+#bridge.tracing=jaeger-otel
+#bridge.tracing.service-name=strimzi-kafka-bridge
```
[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:5]: OK, put back the old one as well + simple description.

File: pom.xml [hunk 11]
```diff
@@ -106,6 +106,9 @@
...
+		<grpc.version>1.41.0</grpc.version>
```
[the remark this answers is not in the document]
Comment (alesj) [reply pom.xml:111]: Afair, gRPC is optional -- as it's just one of the transports that can be used (the most common one).
Hence it needs to be explicitly declared.

File: src/main/java/io/strimzi/kafka/bridge/Application.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java:64]: > If we do not know what and why are we catching, we should probably remove the whole try-catch instead of suppressing the warning.

And have main() throw Exception?

File: src/main/java/io/strimzi/kafka/bridge/converter/DefaultDeserializer.java [hunk 12]
```diff
@@ -19,6 +20,7 @@ public void configure(Map<String, ?> configs, boolean isKey) {
     }
 
     @Override
+    @SuppressFBWarnings("REC_CATCH_EXCEPTION")
```
[the remark this answers is not in the document]
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/converter/DefaultDeserializer.java:23]: Same issue as within Application class ...

File: src/main/java/io/strimzi/kafka/bridge/tracing/TracingUtil.java (hunk 8 above)
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/TracingUtil.java:40]: No, it exists for both -- this is the unique name for your traces / spans (e.g. when you check them in the Jaeger server UI).

File: src/main/java/io/strimzi/kafka/bridge/Application.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java]: To have `main()` throw what exception? The error suggests there is no exception being thrown other than possible RuntimeException which can for sure be thrown out of `main()`.

[the remark this answers is not in the document]
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java]: Then the error is wrong, since there are 3 different non-runtime exceptions being thrown.
Hence probably why we had that all-catch in the first place.

[the remark this answers is not in the document]
Comment (scholzj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java]: Then maybe what you had there was right, but you didn't explained it netiehr in the PR comments nor in the code comments.

[the remark this answers is not in the document]
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java]: I doubt I can explain it better then the error's original docs, from the link posted above, and this one:
* https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#rec-exception-is-caught-when-exception-is-not-thrown-rec-catch-exception

But I fail to see why this approach is better in our case (or for that matter in all cases ...).

File: src/main/java/io/strimzi/kafka/bridge/tracing/TracingConstants.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+ */
+public interface TracingConstants {
+    String JAEGER = "jaeger";
+
+    String JAEGER_OPENTRACING = JAEGER;
+    String JAEGER_OPENTELEMETRY = JAEGER + "-otel";
+
+    String OPENTELEMETRY_SERVICE_NAME_ENV_KEY = "OTEL_SERVICE_NAME";
+    String OPENTELEMETRY_SERVICE_NAME_PROPERTY_KEY = "otel.service.name";
+    String OPENTELEMETRY_TRACES_EXPORTER_KEY = "otel.traces.exporter";
```
[the remark this answers is not in the document]
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/TracingConstants.java:19]: Moved to a class.

File: src/main/java/io/strimzi/kafka/bridge/Application.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java]: I guess I need to look at the actual code.

[the remark this answers is not in the document]
Comment (scholzj) [reply src/main/java/io/strimzi/kafka/bridge/Application.java]: Ok, so, the code throws 3 different exceptions:
* `IOException`
* `MalformedObjectNameException`
* `ParseException`

So the error tells you to do either this:

```java
        } catch (MalformedObjectNameException|IOException|ParseException ex) {
            log.error("Error starting the bridge", ex);
            System.exit(1);
        }
```

eventually something like this if we want to catch the RuntimeException as well:

```java
        } catch (RuntimeException|MalformedObjectNameException|IOException|ParseException ex) {
            log.error("Error starting the bridge", ex);
            System.exit(1);
        }
```

So why not use that? This is what it suggests.

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+    @Override
+    public String envName() {
+        return OPENTELEMETRY_SERVICE_NAME_ENV_KEY;
+    }
+
+    @Override
+    public String serviceName(BridgeConfig config) {
+        String serviceName = System.getenv(envName());
+        if (serviceName == null) {
+            serviceName = System.getProperty(OPENTELEMETRY_SERVICE_NAME_PROPERTY_KEY);
```
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:61]: What is the reason for getting it from the properties suddenly? Do we have 3 way s how to configure the service? In envvar., system propeerty and in the config file?

**Review by scholzj — COMMENTED** [review]
The good news ... the OpenTracing still seems to work fine :-). But I didn't managed to get anything using OTel. When I try to follow the docs PR from @PaulRMellor, it always fails to connect to my Jaeger. Do you have some example configurations for Jaeger Operator and for the agent to play with?

File: config/application.properties [hunk 15]
```diff
@@ -1,7 +1,11 @@
 #Bridge related settings
 bridge.id=my-bridge
-# uncomment the following line to enable Jaeger tracing, check the documentation how to configure the tracer
+# uncomment one the following lines (bridge.tracing) to enable Jaeger tracing, check the documentation how to configure the tracer
+# OpenTracing support
 #bridge.tracing=jaeger
+# OpenTelemetry support
+#bridge.tracing=jaeger-otel
+#bridge.tracing.service-name=strimzi-kafka-bridge
```
[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:8]: Dunno. @ppatierno ? ^

File: config/application.properties [hunk 16]
```diff
@@ -1,7 +1,11 @@
 #Bridge related settings
 bridge.id=my-bridge
-# uncomment the following line to enable Jaeger tracing, check the documentation how to configure the tracer
+# uncomment one the following lines (bridge.tracing) to enable Jaeger tracing, check the documentation how to configure the tracer
+# OpenTracing support
 #bridge.tracing=jaeger
+# OpenTelemetry support
+#bridge.tracing=jaeger-otel
```
[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:7]: Not needed of course.
It was `jaeger` for some unknown reason -- as if there was ever any other option? -- so @ppatierno ? ^^
... and I just followed that unknown reason ...

File: pom.xml [hunk 17]
```diff
@@ -203,6 +206,42 @@
...
+			<groupId>io.opentelemetry</groupId>
+			<artifactId>opentelemetry-exporter-jaeger</artifactId>
+			<version>${opentelemetry-stable.version}</version>
+		</dependency>
+		<!-- Use gRPC as the transport -->
+		<dependency>
+			<groupId>io.grpc</groupId>
+			<artifactId>grpc-netty-shaded</artifactId>
+			<version>${grpc.version}</version>
+		</dependency>
```
[the remark this answers is not in the document]
Comment (alesj) [reply pom.xml:244]: * https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure#jaeger-exporter

`The [Jaeger](https://www.jaegertracing.io/docs/1.21/apis/#protobuf-via-grpc-stable) exporter. This exporter uses gRPC for its communications protocol.`

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java (hunk 14 above)
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:61]: That's how OTel uses this -- env, sys -- but it can be left out.
@ppatierno ^

**Comment by alesj** [comment]
> The good news ... the OpenTracing still seems to work fine :-). But I didn't managed to get anything using OTel. When I try to follow the docs PR from @PaulRMellor, it always fails to connect to my Jaeger. Do you have some example configurations for Jaeger Operator and for the agent to play with?

This is the env vars I have (plus enabled `jaeger-otel` tracing in config)

OTEL_TRACES_EXPORTER=jaeger;JAEGER_SERVICE_NAME=strimzi-kafka-bridge

and it works for me from IDE.

**Comment by scholzj** [comment]
> This is the env vars I have (plus enabled `jaeger-otel` tracing in config)
> 
> OTEL_TRACES_EXPORTER=jaeger;JAEGER_SERVICE_NAME=strimzi-kafka-bridge
> 
> and it works for me from IDE.

Ok, so how do you do it on Kube with Jaeger operator?

File: config/application.properties (hunk 16 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply config/application.properties:7]: Let me rephrase the question ... what it the exact relationship between OTel and Jaeger and how do things like Zipkin fit in?

[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:7]: They don't. :-)
Jaeger is a fixed choice, so imho not a good choice of config value.

**Comment by alesj** [comment]
> > This is the env vars I have (plus enabled `jaeger-otel` tracing in config)
> > OTEL_TRACES_EXPORTER=jaeger;JAEGER_SERVICE_NAME=strimzi-kafka-bridge
> > and it works for me from IDE.
> 
> Ok, so how do you do it on Kube with Jaeger operator?

I think the only thing missing here is the OTEL_EXPORTER_JAEGER_ENDPOINT in k8s.

File: config/application.properties (hunk 16 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply config/application.properties:7]: What does it mean that it is a fixed choice?

[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:7]: It's either `Jaeger` or nothing. No Zipkin, etc ...

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+import io.opentelemetry.instrumentation.kafkaclients.TracingConsumerInterceptor;
+import io.opentelemetry.instrumentation.kafkaclients.TracingProducerInterceptor;
+import io.opentelemetry.sdk.autoconfigure.OpenTelemetrySdkAutoConfiguration;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import io.strimzi.kafka.bridge.config.BridgeConfig;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.kafka.client.producer.KafkaProducerRecord;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.jetbrains.annotations.Nullable;
```
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:30]: I don't think we want to use this. And if we have to, it should be probably declared as dependnecy.

File: config/application.properties (hunk 15 above)
[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:8]: So you could customize your spans / traces names.
It can be fixed to `strimzi-kafka-bridge`.

File: config/application.properties (hunk 16 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply config/application.properties:7]: > It's either Jaeger or nothing. No Zipkin, etc ...

But why? The promise of OpenTelemetry is that it works with many different backends including Jaeger, Zipkin, etc.

File: config/application.properties (hunk 15 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply config/application.properties:8]: I guess the point is, how does it differ from the environment variable and from the system property?

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java (hunk 18 above)
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:30]: Use what? Nullable?

File: config/application.properties (hunk 16 above)
[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:7]: Yeah ... at the end it's that exporter config + right deps on the classpath.
I can add Zipkin, etc ... if you want.

File: config/application.properties (hunk 15 above)
[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:8]: I think we follow the same order -- env, sys, config.
I'm happy to drop any or all of them.

File: config/application.properties (hunk 16 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply config/application.properties:7]: You do not have to add Zipkin. But this clearly impacts the configuration and the initial question.

[the remark this answers is not in the document]
Comment (ppatierno) [reply config/application.properties:7]: the reason was this https://github.com/strimzi/strimzi-kafka-bridge/blob/f96cb3f0958bfcad3c875d54ebb78cedc967a8a7/src/main/java/io/strimzi/kafka/bridge/Application.java#L138 ... this `bridge.tracing` is used to enable tracing and we used jaeger thinking about potential future value with other tracing systems. It's even reflected from the CO when configuring the tracing on the bridge through the STRIMZI_TRACING env var here https://github.com/strimzi/strimzi-kafka-bridge/blob/main/bin/docker/kafka_bridge_config_generator.sh#L4

[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:7]: Ah, OK ... this needs to be fixed then as well.

So I guess `jaeger-otel' is not a good name?
Or, actually, we need a new differentiator? By tracing type not impl ...

[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:7]: > Ah, OK ... this needs to be fixed then as well.

OK, this is fine, it takes the env var ...

File: config/application.properties (hunk 15 above)
[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:8]: As how OTel lookups up config -- env, sys, custom config

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java (hunk 18 above)
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java]: Removed ...

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java (hunk 14 above)
Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:60]: Do we go with all 3 checks or not? Or just exact config property?

File: config/application.properties (hunk 15 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply config/application.properties:8]: So ... the bridge is using the [Vert.x Config](https://vertx.io/docs/vertx-config/java/) for getting the configuration.
As you can see [here ](https://github.com/strimzi/strimzi-kafka-bridge/blob/main/src/main/java/io/strimzi/kafka/bridge/Application.java#L95) we provides two stores for configuration: file (the application.properties) and env vars. Any env var value overrides the same parameter specified in the properties file (because how stores are added matters and in our case is `fileStore` first and then `envStore` as explained [here](https://vertx.io/docs/vertx-config/java/#_overloading_rules)).

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java (hunk 14 above)
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:60]: @alesj I didn't get the question for me? Isn't it the code you added. If you want to know how bridge configuration works, take a look at the previous comment about using [Vert.x Config](https://vertx.io/docs/vertx-config/java/#_overloading_rules) so in our case (due to stores order) env vars first and application.properties file ... but no system properties at all.

Comment (alesj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:60]: OK, as "said" just before, I'll remove sys prop lookup.

File: config/application.properties [hunk 19]
```diff
@@ -1,7 +1,11 @@
 #Bridge related settings
 bridge.id=my-bridge
-# uncomment the following line to enable Jaeger tracing, check the documentation how to configure the tracer
+# uncomment one the following lines (bridge.tracing) to enable Jaeger tracing, check the documentation how to configure the tracer
+# OpenTracing support
 #bridge.tracing=jaeger
+# OpenTelemetry support
+bridge.tracing=opentelemetry
+bridge.tracing.service-name=strimzi-kafka-bridge
```
[the remark this answers is not in the document]
Comment (alesj) [reply config/application.properties:8]: Ah, testing ... and forgot to comment it back. Fixed.

File: bin/docker/kafka_bridge_run.sh [hunk 20]
```diff
@@ -68,6 +68,11 @@ if [ "$FIPS_MODE" = "disabled" ]; then
     export JAVA_OPTS="${JAVA_OPTS} -Dcom.redhat.fips=false"
 fi
 
+# enabling OpenTelemetry with Jaeger by default
+if [ -n "$OTEL_SERVICE_NAME" ] && [ -z "$OTEL_TRACES_EXPORTER" ]; then
+  export OTEL_TRACES_EXPORTER="jaeger"
+fi
```
[the remark this answers is not in the document]
Comment (alesj) [reply bin/docker/kafka_bridge_run.sh:74]: Imo no need for `STRIMZI_TRACING` gates.
I'll remove this snippet, and just leave similar one in bin/kafka_bridge_run.sh

File: bin/kafka_bridge_run.sh [hunk 21]
```diff
@@ -13,4 +13,9 @@ fi
 # Make sure that we use /dev/urandom
 JAVA_OPTS="${JAVA_OPTS} -Dvertx.cacheDirBase=/tmp/vertx-cache -Djava.security.egd=file:/dev/./urandom"
 
+# enabling OpenTelemetry with Jaeger by default
+if [ -n "$OTEL_SERVICE_NAME" ] && [ -z "$OTEL_TRACES_EXPORTER" ]; then
```
[the remark this answers is not in the document]
Comment (alesj) [reply bin/kafka_bridge_run.sh:17]: Removed.

File: pom.xml [hunk 22]
```diff
@@ -491,6 +546,11 @@
 									<ignoredUnusedDeclaredDependency>io.jaegertracing:jaeger-client</ignoredUnusedDeclaredDependency>
 									<ignoredUnusedDeclaredDependency>org.yaml:snakeyaml</ignoredUnusedDeclaredDependency> <!-- CVE override -->
 									<ignoredUnusedDeclaredDependency>org.apache.tomcat.embed:tomcat-embed-core</ignoredUnusedDeclaredDependency>  <!-- CVE override -->
+									<!-- OpenTelemetry - used via classpath configuration -->
+									<ignoredUnusedDeclaredDependency>io.opentelemetry:opentelemetry-sdk-trace</ignoredUnusedDeclaredDependency>  <!-- CVE override -->
+									<ignoredUnusedDeclaredDependency>io.opentelemetry.instrumentation:opentelemetry-kafka-clients-common</ignoredUnusedDeclaredDependency>  <!-- CVE override -->
+									<ignoredUnusedDeclaredDependency>io.opentelemetry:opentelemetry-exporter-jaeger</ignoredUnusedDeclaredDependency>  <!-- CVE override -->
+									<ignoredUnusedDeclaredDependency>io.grpc:grpc-netty-shaded</ignoredUnusedDeclaredDependency>  <!-- CVE override -->
```
[the remark this answers is not in the document]
Comment (alesj) [reply pom.xml:553]: Ah, copy paste ...

**Comment by alesj** [comment]
Closed in favor of a new PR: https://github.com/strimzi/strimzi-kafka-bridge/pull/633
